sdk/search/Azure.Search.Documents/src/Models/SearchContinuationToken.cs

```diff
@@ -0,0 +1,104 @@
+
+        private const string NextLinkName = "nextLink";
+        private static readonly JsonEncodedText s_nextLinkEncodedName = JsonEncodedText.Encode(NextLinkName);
+
+        private const string NextPageParametersName = "nextPageParameters";
+        private static readonly JsonEncodedText s_nextPageParametersEncodedName = JsonEncodedText.Encode(NextPageParametersName);
+
+        /// <summary>
+        /// Creates a durable, opaque continuation token that can be provided
+        /// to users.
+        /// </summary>
+        /// <param name="nextPageUri">
+        /// URI of additional results when making GET requests.
+        /// </param>
+        /// <param name="nextPageOptions">
+        /// <see cref="SearchOptions"/> for additional results when making POST
+        /// requests.
+        /// </param>
+        /// <param name="apiVersion">Versioning discriminator.</param>
+        /// <returns>A continuation token.</returns>
+        public static string Serialize(
+            Uri nextPageUri,
+            SearchOptions nextPageOptions,
+            SearchClientOptions.ServiceVersion apiVersion = SearchClientOptions.ContinuationTokenVersion)
+        {
+            // There's no continuation token if there's no next page
+            if (nextPageUri == null || nextPageOptions == null)
+            {
+                return null;
+            }
+            using MemoryStream buffer = new MemoryStream();
+            using (Utf8JsonWriter writer = new Utf8JsonWriter(buffer))
+            {
+                writer.WriteStartObject();
+                writer.WriteString(s_apiVersionEncodedName, apiVersion.ToVersionString());
+                writer.WriteString(s_nextLinkEncodedName, nextPageUri.ToString());
+                writer.WritePropertyName(s_nextPageParametersEncodedName);
+                writer.WriteObjectValue(nextPageOptions);
+                writer.WriteEndObject();
+            }
+            return Convert.ToBase64String(buffer.ToArray());
+        }
+
+        /// <summary>
+        /// Parses a continuation token and returns the corresponding to the
+        /// next page's <see cref="SearchOptions"/>.
+        /// </summary>
+        /// <param name="continuationToken">
+        /// The serialized continuation token.
+        /// </param>
+        /// <returns>The continuation token's next page options.</returns>
+        public static SearchOptions Deserialize(string continuationToken)
+        {
+            Argument.AssertNotNullOrEmpty(continuationToken, nameof(continuationToken));
+            byte[] decoded = Convert.FromBase64String(continuationToken);
+            try
+            {
+                using JsonDocument json = JsonDocument.Parse(decoded);
+                if (json.RootElement.ValueKind == JsonValueKind.Object &&
+                    json.RootElement.TryGetProperty(ApiVersionName, out JsonElement apiVersion) &&
```

Comment:
`TryGetProperty` is slower. Each call is O(n) where n is the number of child elements. Instead, follow the same pattern as the generated serialize and use `EnumerateObject()`.

Reply:
I don't quite understand this, EnumerateObject does N checks N times so it's same O(N*N) as `TryGetProperty` or am I missing something?

Reply:
No, it loops through each property. You'd use a switch or if condition to handle each property name. You do this in your generator, and we do it in our other non-generated services already. Each call to TryGetProperty is O(n), so three calls triples the time (still technically O(n), but big O notation isn't an absolute order of perf). If you need to get more than one property, you use `EnumerateObject`.

Reply:
I still don't understand.

```
foreach (prop in EnumerateObject())
{
	if (prop.name == "a") {}
	if (prop.name == "b") {}
	if (prop.name == "c") {}
}
```

is still a N*N, as it does N loops with N comparisons in each


Same as 
```
TryGetProperty("a")
TryGetProperty("b")
TryGetProperty("c")
```
where every call does 1 loops with N comparisons in each.

Reply:
The first is O(n). If you want to count the comparison, it's O(nm). The relative performance of the latter would be O(3nm). The latter also allocates (even though on the stack, it's still an extra allocation) for each enumerate it creates internally. We've discussed this as a team previously and prefer `EnumerateObject`. You're doing it in all the deserialization code right now. Why is this an issue all of a sudden?

Reply:
Given that it will only have 3 properties, I'd prefer to leave it written like this because I think it's cleaner than turning it into a state machine and O(27) doesn't bother me for niche scenarios we don't really want customers using.  I will also be adding an additional `ValueKind` check here after pinging @ahsonkhan about it (the `TryGet*` methods will still throw if the `ValueKind` is incorrect...  I didn't grasp that initially).

I can change it to use `EnumerateObject` if everyone thinks this a poor choice.

Reply:
It's not an issue, I'm fine with both implementations I just don't understand the reasoning and I want to.

Reply:
TryGetProperty seems to be faster:

```

|         Method |     Mean |    Error |   StdDev |
|--------------- |---------:|---------:|---------:|
| TryGetProperty | 499.9 ns | 13.33 ns | 39.30 ns |
|           Loop | 738.6 ns | 22.75 ns | 67.08 ns |
```

```
using System;
using System.Collections.Generic;
using System.Diagnostics;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Text.Json;

namespace activity_bench
{
    public class JElementBench
    {
        public static readonly JsonElement e = JsonDocument.Parse(@"{
           ""a"":""a"",
           ""b"":""b"",
           ""c"":""c"",
           ""d"":""d"",
           ""f"":""f""
            }").RootElement;

        [Benchmark]
        public void TryGetProperty()
        {
            e.TryGetProperty("a", out var a);
            e.TryGetProperty("b", out var b);
            e.TryGetProperty("c", out var c);
            e.TryGetProperty("d", out var d);
            e.TryGetProperty("f", out var f);
        }

        [Benchmark]
        public void Loop()
        {
            JsonElement a;
            JsonElement b;
            JsonElement c;
            JsonElement d;
            JsonElement f;

            foreach (var prop in e.EnumerateObject())
            {
                if (prop.Name == "a") a = prop.Value;
                if (prop.Name == "b") a = prop.Value;
                if (prop.Name == "c") a = prop.Value;
                if (prop.Name == "d") a = prop.Value;
                if (prop.Name == "f") a = prop.Value;
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            var summary = BenchmarkRunner.Run<JElementBench>();
        }
    }
}

```